qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1773 @@
+            idx = idx_or_label
+        else:
+            idx = self._ados.idx(idx_or_label)
+        return Qobj(self._ado_state[idx, :].T, dims=self.rho.dims)
+
+
+class HEOMSolver:
+    """
+    HEOM solver that supports multiple baths.
+
+    The baths must be all either bosonic or fermionic baths.
+
+    Parameters
+    ----------
+    H_sys : QObj, QobjEvo or a list
+        The system Hamiltonian or Liouvillian specified as either a
+        :obj:`Qobj`, a :obj:`QobjEvo`, or a list of elements that may
+        be converted to a :obj:`ObjEvo`.
+
+    baths : Bath or list of Bath
+        A :obj:`Bath` containing the exponents of the expansion of the
+        bath correlation funcion and their associated coefficients
+        and coupling operators.
+
+        If multiple baths are given, they must all be either fermionic
+        or bosonic baths.
+
+    max_depth : int
+        The maximum depth of the heirarchy (i.e. the maximum number of bath
+        exponent "excitations" to retain).
+
+    options : :class:`qutip.solver.Options`
+        Generic solver options. If set to None the default options will be
+        used.
+
+    Attributes
+    ----------
+    ados : :obj:`HierarchyADOs`
+        The description of the hierarchy constructed from the given bath
+        and maximum depth.
+    """
+    def __init__(self, H_sys, bath, max_depth, options=None):
+        self.H_sys = self._convert_h_sys(H_sys)
+        self.options = Options() if options is None else options
+        self._is_timedep = isinstance(self.H_sys, QobjEvo)
+        self._H0 = self.H_sys.to_list()[0] if self._is_timedep else self.H_sys
+        self._is_hamiltonian = self._H0.type == "oper"
+        self._L0 = liouvillian(self._H0) if self._is_hamiltonian else self._H0
+
+        self._sys_shape = (
+            self._H0.shape[0] if self._is_hamiltonian
+            else int(np.sqrt(self._H0.shape[0]))
+        )
+        self._sup_shape = self._L0.shape[0]
+        self._sys_dims = (
+            self._H0.dims if self._is_hamiltonian
+            else self._H0.dims[0]
+        )
+
+        self.ados = HierarchyADOs(
+            self._combine_bath_exponents(bath), max_depth,
+        )
+        self._n_ados = len(self.ados.labels)
+        self._n_exponents = len(self.ados.exponents)
+
+        # pre-calculate identity matrix required by _grad_n
+        self._sId = fast_identity(self._sup_shape)
+
+        # pre-calculate superoperators required by _grad_prev and _grad_next:
+        Qs = [exp.Q for exp in self.ados.exponents]
+        self._spreQ = [spre(op).data for op in Qs]
+        self._spostQ = [spost(op).data for op in Qs]
+        self._s_pre_minus_post_Q = [
+            self._spreQ[k] - self._spostQ[k] for k in range(self._n_exponents)
+        ]
+        self._s_pre_plus_post_Q = [
+            self._spreQ[k] + self._spostQ[k] for k in range(self._n_exponents)
+        ]
+
+        self.progress_bar = BaseProgressBar()
```

Comment:
The solving is probably slow and so we should not hide the progress bar here.
It should at least be mentioned in the doc-string.

Reply:
In v4, it's a parameter in the `..solve` function and in v5 it will be an option.

Reply:
I've added it as a parameter in the v4 style for now. Thanks!